### PR TITLE
Data links

### PR DESCRIPTION
**Pull request by torkelo**
Part of enhanced drilldown links work #15130 

* [x] WIP Panel links editor with just title & URL 
* [x] Open in new tab switch
* [x] Migration for old option schema to variables 
  * [x] prior v5 links migrattion
  * [x] dashUri migration
* [x]  Update link srv to interpolate the new variables 
* [x] Interpolate `$__series` and `$__datapointTs` variables (https://github.com/grafana/grafana/pull/17328)
* [x] Context menu - add details about data point clicked (ts, series name)
* [x] Separate panel links and visualisation links (#17355)
* [x] Bring `LinkSrv.getPanelLinkAnchorInfo` back for plugins compatibility 
* [x] Rename Drilldown links to Data links
* [x] Update docs

### CODE REVIEW

File: public/app/features/panel/PanelLinksEditor/PanelLinksEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  return (
+    <PanelOptionsGroup title="Drilldown links" onAdd={onAdd}>
+      {value.map((link, index) => (
+        <LinkEditor key={index.toString()} index={index} value={link} onChange={onLinkChanged} onRemove={onRemove} />
+      ))}
+    </PanelOptionsGroup>
+  );
+});
+
+export interface LinkEditor {
```
Comment (dprokop) [inline public/app/features/panel/PanelLinksEditor/PanelLinksEditor.tsx]:
```suggestion
export interface LinkEditorProps {
```

File: public/app/features/panel/PanelLinksEditor/PanelLinksEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  index: number;
+  value: PanelDrillDownLink;
+  onChange: (index: number, link: PanelDrillDownLink) => void;
+  onRemove: (link: PanelDrillDownLink) => void;
+}
+
+export const LinkEditor: FC<LinkEditor> = React.memo(({ index, value, onChange, onRemove }) => {
```
Comment (dprokop) [inline public/app/features/panel/PanelLinksEditor/PanelLinksEditor.tsx]:
```suggestion
export const LinkEditor: FC<LinkEditorProps> = React.memo(({ index, value, onChange, onRemove }) => {
```

**Comment by dprokop** [comment]
@torkelo renamed this PR as it started to become sth more than planned. Let's treat it as a feature branch for drilldown work

File: public/app/features/panel/panellinks/link_srv.ts [hunk 3]
```diff
@@ -71,45 +94,89 @@ export class LinkSrv {
...
+    const timeRange = this.timeSrv.timeRange();
+    const targetAbsoluteTimeRange = transformAbsoluteTimeRange(timeRange, DataLinkDataPointRangeFactor);
+    const dataPointRangeQuery = toUrlParams({
+      from: dateTime(valueTime)
+        .subtract(targetAbsoluteTimeRange)
+        .valueOf(),
+      to: dateTime(valueTime)
+        .add(targetAbsoluteTimeRange)
+        .valueOf(),
+    });
```
Comment (torkelo) [inline public/app/features/panel/panellinks/link_srv.ts]: Think there is a missunderstanding here, the ${__value_time} should only interpolate to the time, not from & to. If the user wants that they will need to set &from=${__value_time}&to=${__value_time}

Comment (dprokop) [reply public/app/features/panel/panellinks/link_srv.ts]: Fixed

File: public/app/features/panel/panellinks/link_srv.ts [hunk 4]
```diff
@@ -20,96 +80,92 @@ export class LinkSrv {
...
-    const params = {};
-
-    if (link.keepTime) {
-      const range = this.timeSrv.timeRangeForUrl();
-      params['from'] = range.from;
-      params['to'] = range.to;
-    }
+  getDataPointVars = (seriesName: string, valueTime: DateTime) => {
+    const valueTimeQuery = toUrlParams({
+      time: dateTime(valueTime).valueOf(),
```
Comment (torkelo) [inline public/app/features/panel/panellinks/link_srv.ts]: this looks wrong, valueTime is already a DateTime here, should be safe just to call valueOf() directly?

File: public/app/features/panel/panellinks/link_srv.ts [hunk 5]
```diff
@@ -20,96 +80,92 @@ export class LinkSrv {
...
+      },
+    };
+  };
+
+  getDataLinkUIModel = (link: DataLink, scopedVars: ScopedVars, dataPoint?: LinkDataPoint) => {
+    const params: KeyValue = {};
+    const timeRangeUrl = toUrlParams(this.timeSrv.timeRangeForUrl());
+
+    const info: LinkModel = {
+      url: link.url,
```
Comment (torkelo) [inline public/app/features/panel/panellinks/link_srv.ts]: seems to be some typing issues here, in the old panels the expected return prop is href, so getting type errors due to that.

**Review by torkelo — COMMENTED** [review]
Some remaining issues.

* The return link UI model (LinkModel in LinkSrv) has changed (in a breaking way). Was before href, think we need to change this back to href for backwards compatible reasons. Example of this can be found in singlestat (currently generated type errors due to this as Tobias added types there). 
* The $__value_time still interpolates incorrectly, interpolates as "time=213211122" instead of just the 
epoch number. 
http://localhost:3000/d/5SdHCadmz/panel-tests-graph?orgId=1&from=${__value_time}
renders as: 
http://localhost:3000/d/5SdHCadmz/panel-tests-graph?orgId=1&from=time=1560946393452

**Pull request merged**